python/ray/train/_internal/session.py

```diff
@@ -424,7 +434,71 @@ def _set_checkpoint_uri(self, uri: str):
         Args:
             uri: URI to the location where next checkpoint should be saved.
         """
-        self.checkpoint_uri = uri
+        if _use_storage_context():
+            # TODO(justinvyu): Fix up the name.
+            self.storage.current_checkpoint_id = uri
+        else:
+            self.legacy_checkpoint_uri = uri
+
+    def new_checkpoint(self, checkpoint: NewCheckpoint):
+        import pyarrow.fs
+
+        assert isinstance(checkpoint, NewCheckpoint)
+        assert isinstance(checkpoint.filesystem, pyarrow.fs.LocalFileSystem)
+
+        # Upload checkpoint files.
+        print(
+            "Uploading checkpoint files to storage path:"
+            "\n{source}\n{destination}\n{source_fs}\n{dest_fs}".format(
+                source=checkpoint.path,
+                destination=self.storage.checkpoint_fs_path,
+                source_fs=checkpoint.filesystem,
+                dest_fs=self.storage.storage_filesystem,
+            )
+        )
+        self.storage.storage_filesystem.create_dir(self.storage.checkpoint_fs_path)
+        pyarrow.fs.copy_files(
+            source=checkpoint.path,
+            destination=self.storage.checkpoint_fs_path,
+            source_filesystem=checkpoint.filesystem,
+            destination_filesystem=self.storage.storage_filesystem,
+        )
+        # Delete local checkpoint files.
+        checkpoint.filesystem.delete_dir(checkpoint.path)
+
+        # Report the uploaded checkpoint location for internal book-keeping.
+        checkpoint_to_report = NewCheckpoint(
+            filesystem=self.storage.storage_filesystem,
+            path=self.storage.checkpoint_fs_path,
+        )
+
+        metadata = self._auto_fill_checkpoint_metrics({})
+
+        # Save the rank of the worker that created this checkpoint.
+        metadata.update({CHECKPOINT_RANK_KEY: self.world_rank})
+        checkpoint_to_report.set_metadata(metadata)
```

Comment:
Hmm we shouldn't set this on the actual checkpoint right?

Reply:
This is the internal metadata that I want to attach to a checkpoint. We should be able to know the metrics that were reported with a checkpoint (useful for tracking top k checkpoints).

We currently have a few metadata concepts that should be cleaned up:
1. The old `.metadata.pkl` file. This contained the checkpoint class information and was used for framework-specific checkpoints, but it's not needed anymore.
2. The new `Checkpoint.set_metadata`. User-facing metadata.
3. Internal metadata that was previously stored in the `_TrackedCheckpoint`.  This is pretty much just the metrics reported with the checkpoint. This PR adds it as part of the user-facing metadata.
    - **Q: Would it be better to separate 2 and 3? Ex: we can dump internal metadata in a separate `.internal_metadata.json` file.**
4. Internal metadata that is stored in the `.tune_metadata`. This consists of trainable state and can probably be removed in the new path.

Reply:
Those metadata are available in the Result object though as metrics right? It's weird to put a couple metrics in Checkpoint, which should be only opaque user data.

Afaik, the result metadata is stored in the dir separately already.

Reply:
For clarity, I believe the structure we want is
```
Result
   metrics (incl tune internal metadata/metrics)
   Checkpoint 
       user data
       user metadata 
```

Reply:
Yep, the full set of results are saved in the trial dir, but I think a checkpoint should still know the set of metrics that was reported along with it. This info is currently stored in a few places:
- The `experiment-state.json` file holds onto these `_TrackedCheckpoint` objects, which hold onto "metadata" which is the metrics reported with the checkpoint. This is at the experiment level directory, and we're trying to move away from storing data in that file.
- The `.tune_metadata` stores certain parts of state, such as the training iteration, and it directly within the checkpoint folder.

I'm thinking we can do:

```
trial_dir/
    result.json -- all metrics
    checkpoint_0/
        .internal_metadata.json <-- metrics & other metadata we populate, associated with the checkpoint
        .user_metadata.json
        user data ...
```

Reply:
It's redundant with result.json etc right? In the result.json, we can store "checkpoint_number": "00001" etc. to connect which checkpoint should be associated with which metric.

Anyways, we shouldn't discuss this in the scope of 2.7--- it's a new feature that we can tack on later once we finish the GA APIs as specced out.